File: tests/journal.ext/error_loop.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2022 all rights reserved


def test():
    """
    Verify that repeated use of the same channel instance works correctly
    """
    # get the channel
    from journal.ext.journal import Error as error
    # and the trash can
    from journal.ext.journal import Trash as trash

    # make a channel
    channel = error(name="test.journal.error")
    # make it non-fatal
    channel.fatal = False
    # send the output to trash
    channel.device = trash()

    # a few times
    for _ in range(10):
        # inject
        channel.log("hello world!")

    # all done
    return


# main
if __name__ == "__main__":
    # run the test
    test()


# end of file

File: tests/journal.lib/error_loop.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2023 all rights reserved


// get the journal
#include <pyre/journal.h>
// support
#include <cassert>


// type aliases
using myerror_t = pyre::journal::error_t;
using trash_t = pyre::journal::trash_t;


// verify that repeated injections work correctly
int main() {
    // make an error channel
    myerror_t channel("tests.journal.error");

    // send the output to the trash
    channel.device<trash_t>();

    // inject repeatedly
    for (auto i=0; i<10; ++i) {
        // carefully
        try {
            // inject something
            channel << "i: " << i << pyre::journal::endl;
            // errors are fatal by default, so we shouldn't be able to get here
            throw std::logic_error("unreachable");
            // if all goes well
        }  catch (const myerror_t::exception_type & error) {
            // make sure the reason was recorded correctly
            assert (error.what() == channel.name() + myerror_t::string_type(": application error"));
        }
    }

    // all done
    return 0;
}


// end of file
